File: Design Snake Game.java
/*
class SnakeGame:

    def __init__(self, width: int, height: int, food: List[List[int]]):
        self.width = width
        self.height = height
        self.snake = list()
        self.head = [0,0]
        self.snake.append(self.head)
        self.foodlist = collections.deque()
        
        for i in food:
            self.foodlist.append(i)

    def move(self, direction: str) -> int:
        if direction == "U":
            self.head[0] -= 1
        
        if direction == "D":
            self.head[0] += 1
        
        if direction == "L":
            self.head[1] -= 1
        
        if direction == "R":
            self.head[1] += 1
        
        # boundary
        if (self.head[0] < 0 or self.head[0] >= self.height or self.head[1] < 0 or self.head[1] >= self.width):
            return -1
        
        # touches itself
        for i in range(1, len(self.snake)):                 // can use a hashset here as searching will be O(1) and then entire program would be O(1)
            cur = self.snake[i]
            if self.head[0] == cur[0] and self.head[1] == cur[1]:
                return -1
        
        # food 
        if len(self.foodlist) > 0:
            food = self.foodlist[0]
            if food[0] == self.head[0] and food[1] == self.head[1]:
                self.snake.append(self.foodlist.popleft())
                return len(self.snake) - 1
        
        self.snake.pop(0)                   # use linkedlist though because in a normal list popleft and append                                             # from right will cause O(n) operation
        self.snake.append([]+self.head) 
        return len(self.snake) - 1
*/

// Time - O(width*height) it comes from for loop which checks whether snake touches itself and snake can be as long as matrix
// Space - O(width*height) for storing snake and food
// logic - stored snake and food in linkedlist and the head of linkedlist will be tail of snake. Everytime snake eats food I am adding that food
// coordinates to snake's head i.e tail of linkedlist. Everytime a snake moves I am removing from tail i.e head of linkedlist and adding next
// coordinate to head of snake i.e tail of linkedlist. Can also use a queue here but then all snake body coordinates will have to be present in 
// hashset as while checking if snakes touches itself we cant iterate over queue
class SnakeGame {

    LinkedList<int[]> snake;
    LinkedList<int[]> foodList;
    int[] head;
    int width, height;
    public SnakeGame(int width, int height, int[][] food) {
        head = new int[] {0,0};
        snake = new LinkedList<>();
        foodList = new LinkedList<>(Arrays.asList(food));
        snake.add(head);
        this.width = width;
        this.height = height;
    }
    
    public int move(String direction) {
        if (direction.equals("U"))
            head[0]--;
        
        if (direction.equals("D"))
            head[0]++;
        
        if (direction.equals("R"))
            head[1]++;
        
        if (direction.equals("L"))
            head[1]--;
        
        // if snake touchese boundary
        if (head[0] < 0 || head[0] >= height || head[1] < 0 || head[1] >= width)
            return -1;
/*
----
 ^ |
 |_|          even this test case will pass as after move the tail we have is no longer valid and hence we should 
            start looking from index 1
*/
        // touches itself
        for (int i=1; i<snake.size(); i++){
            int[] body = snake.get(i);
            if (head[0] == body[0] && head[1] == body[1])
                return -1;
        }
        
        if (foodList.size() > 0){
            int[] food_index = foodList.get(0);
            if (head[0] == food_index[0] && head[1] == food_index[1]){
                snake.add(foodList.remove());
                return snake.size() - 1;        // -1 because initially we add head which doesnt count                                                           //   towawrds point
            }
                               
        }
        snake.remove();
        snake.add(new int[] {head[0], head[1]});
        return snake.size() - 1;  
    }
}

File: design-snake-game.py
# Time Complexity: O(1), dispaly size will be limited
# Space Complexity: O(m*n)
# class SnakeGame:
#   def __init__(self,height,width,food):
#     from collections import deque
#     self.snake=deque()
#     self.h=height
#     self.w=width
#     self.food=food
#     self.head=[0,0]
#     self.snake.append(self.head)
#     self.idx=0
#   def move(self,direction):
#     if direction=='L': self.head[1]-=1
#     elif direction=='R': self.head[1]+=1 
#     elif direction=='U': self.head[0]-=1
#     elif direction=='D': self.head[0]+=1
#     # hitting the borders
#     if self.head[0]<0 or self.head[1]<0 or self.head[0]==self.h or self.head[1]==self.w: return -1
#     # hitting the snake
#     for i in range(1,len(self.snake)):
#       if self.snake[i][0]==self.head[0] and self.snake[i][1]==self.head[1]: return -1
#     # food move
#     if self.idx<len(self.food) and self.head[0]==self.food[self.idx][0] and self.head[1]==self.food[self.idx][1]:
#       self.snake.append(self.head.copy())
#       self.idx+=1
#       return len(self.snake)-1
#     # normal move
#     self.snake.append(self.head.copy())
#     self.snake.popleft()
#     return len(self.snake)-1
# snake=SnakeGame(2,3,[[1,2],[0,1]])
# print(snake,snake.move("R"),snake.move("D"),snake.move("R"),snake.move("U"),snake.move("L"),snake.move("U"))
# Time Complexity: O(1), dispaly size will be limited
# Space Complexity: O(m*n)
class SnakeGame:
  def __init__(self,height,width,food):
    from collections import deque
    self.snake=deque()
    self.h=height
    self.w=width
    self.food=food
    self.head=[0,0]
    self.board=[[False for j in range(self.w)] for i in range(self.h)]
    self.snake.append(self.head)
    self.idx=0
  def move(self,direction):
    if direction=='L': self.head[1]-=1
    elif direction=='R': self.head[1]+=1 
    elif direction=='U': self.head[0]-=1
    elif direction=='D': self.head[0]+=1
    # hitting the borders
    if self.head[0]<0 or self.head[1]<0 or self.head[0]==self.h or self.head[1]==self.w: return -1
    # hitting the snake
    if self.board[self.head[0]][self.head[1]]==True: return -1
    # food move
    if self.idx<len(self.food) and self.head[0]==self.food[self.idx][0] and self.head[1]==self.food[self.idx][1]:
      self.snake.append(self.head.copy())
      self.board[self.head[0]][self.head[1]]=True
      self.idx+=1
      return len(self.snake)-1
    # normal move
    self.snake.append(self.head.copy())
    self.board[self.head[0]][self.head[1]]=True
    self.snake.popleft()
    temp=self.snake[0]
    self.board[temp[0]][temp[1]]=False
    return len(self.snake)-1
snake=SnakeGame(2,3,[[1,2],[0,1]])
print(snake,snake.move("R"),snake.move("D"),snake.move("R"),snake.move("U"),snake.move("L"),snake.move("U"))